### ISSUE TITLE
Rule .each will fail for non arrays

### ISSUE THREAD

**Issue by alexandru-calinoiu** [dry-rb/dry-validation#567]
**Describe the bug**

Array rules will fail if the user feeds non array data in the fields

**To Reproduce**

Give a rule like this:

```ruby
rule(:steps).each do
   key.failure(:member_exists?) unless member_exists?(value[:id])
end
```
running `Contract.call(steps: '42')` will throw an exception.

**Expected behavior**

Should consider that the value might not be an array and don't run the validations.

**Your environment**

- Affects my production application: NO
- Ruby version: 2.6.2
- OS: Linux

**Comment by alexandru-calinoiu** [comment]
Hm, loooking at the code https://github.com/dry-rb/dry-validation/blob/29d4aff9e3f14c3aae8244a0123481746fd51333/spec/integration/contract/class_interface/rule/each_spec.rb#L37 it should not run the rule if it does not passes schema checks, but for a reason it does not do that in my case.

**Comment by alexandru-calinoiu** [comment]
Here is a sample contract that fails:

```
          contract_class = Class.new(Dry::Validation::Contract) do
            def self.name
              'RecContract'
            end

            params do
              required(:steps).array(:hash)
            end

            rule(:steps).each do
              key.failure(:member_exists?) unless member_exists?(value[:id])
            end
          end

          contract_class.new.(params)
```

**Comment by alexandru-calinoiu** [comment]
I've drafted a pr with a test that shows the issue and a possible fix.

**Comment by alexandru-calinoiu** [comment]
Unfortunately this is not fixed in the 1.2.1, I get the same error with the sample above :(

**Comment by solnic** [comment]
@alexandru-calinoiu the spec you wrote in #568 passes. Please provide another test that fails.
